core/src/main/java/software/amazon/awssdk/interceptor/ExecutionInterceptorChain.java

```diff
@@ -0,0 +1,162 @@
+        interceptors.forEach(i -> i.beforeTransmission(context, executionAttributes));
+    }
+
+    public void afterTransmission(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        reverseForEach(i -> i.afterTransmission(context, executionAttributes));
+    }
+
+    public InterceptorContext modifyHttpResponse(InterceptorContext context,
+                                                 ExecutionAttributes executionAttributes) {
+        InterceptorContext result = context;
+        for (int i = interceptors.size() - 1; i >= 0; i--) {
+            SdkHttpFullResponse interceptorResult = interceptors.get(i).modifyHttpResponse(result, executionAttributes);
+            validateInterceptorResult(result.httpResponse(), interceptorResult, interceptors.get(i), "modifyHttpResponse");
+
+            result = result.modify(b -> b.httpResponse(interceptorResult));
+        }
+        return result;
+    }
+
+    public void beforeUnmarshalling(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        reverseForEach(i -> i.beforeUnmarshalling(context, executionAttributes));
+    }
+
+    public void afterUnmarshalling(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        reverseForEach(i -> i.afterUnmarshalling(context, executionAttributes));
+    }
+
+    public InterceptorContext modifyResponse(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        InterceptorContext result = context;
+        for (int i = interceptors.size() - 1; i >= 0; i--) {
+            SdkResponse interceptorResult = interceptors.get(i).modifyResponse(result, executionAttributes);
+            validateInterceptorResult(result.response(), interceptorResult, interceptors.get(i), "modifyResponse");
+
+            result = result.modify(b -> b.response(interceptorResult));
+        }
+        return result;
+    }
+
+    public void afterExecution(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        reverseForEach(i -> i.afterExecution(context, executionAttributes));
+    }
+
+    public void onExecutionFailure(DefaultFailedExecutionInterceptorContext context, ExecutionAttributes executionAttributes) {
+        interceptors.forEach(i -> i.onExecutionFailure(context, executionAttributes));
+    }
+
+    /**
+     * Validate the result of calling an interceptor method that is attempting to modify the message to make sure its result is
+     * valid.
+     */
+    private void validateInterceptorResult(Object originalMessage, Object newMessage,
+                                           ExecutionInterceptor interceptor, String methodName) {
+        if (originalMessage != newMessage) {
+            LOG.debug(() -> "Interceptor '" + interceptor + "' modified the message with its " + methodName + " method.");
+            LOG.trace(() -> "Old Message: " + originalMessage);
```

Comment:
Nit: why "message" and not "object"? Also should these be in the same log statement? What if there's interleaving log messages being written by another thread?